### PR TITLE
rgw: segment fault when shard id out of range

### PR DESCRIPTION
**Pull request by mingshuaiwang**
Signed-off-by: mingshuaiwang <736808191@qq.com>

### CODE REVIEW

**Comment by mingshuaiwang** [comment]
hi @yehudasa,  can you help to review this pr?

**Comment by yehudasa** [comment]
@mingshuaiwang looks ok. Can you open a ceph tracker issue for this problem and reference it using `Fixes:` in the commit message? Thanks

**Comment by mingshuaiwang** [comment]
hi @yehudasa,
this is the tracker issue: [fix: rgw crashed caused by shard id out of range when listing data log](http://tracker.ceph.com/issues/19732)
is that ok?

**Comment by mattbenjamin** [comment]
@mingshuaiwang I think so;  you should amend the commit message to add:

"Fixes: http://tracker.ceph.com/issues/19732"

File: src/rgw/rgw_bucket.cc [hunk 1]
```diff
@@ -1730,6 +1730,8 @@ int RGWDataChangesLog::list_entries(int shard, const real_time& start_time, cons
 				    const string& marker,
 				    string *out_marker,
 				    bool *truncated) {
+  if (shard > num_shards)
+    return -EINVAL;
```
Comment (cbodley) [inline src/rgw/rgw_bucket.cc]: `if (shard >= num_shards)`

**Comment by mattbenjamin** [comment]
@cbodley  thinks this may be reproducible by an ordinary S3 user, which...

**Comment by cbodley** [comment]
the user would need the `datalog` read cap to hit this

**Comment by mingshuaiwang** [comment]
hi, @mattbenjamin I have changed the commit message

**Comment by mattbenjamin** [comment]
@cbodley cool

**Comment by mingshuaiwang** [comment]
@mattbenjamin Got it， thank you~

File: src/rgw/rgw_bucket.cc (hunk 1 above)
Comment (cbodley) [inline src/rgw/rgw_bucket.cc]: ping @mingshuaiwang, can you fix this check please?

Comment (mingshuaiwang) [inline src/rgw/rgw_bucket.cc]: @cbodley  sorry, my bad,  fixed~

**Comment by cbodley** [comment]
passed rgw suite: http://pulpito.ceph.com/cbodley-2017-05-22_11:00:27-rgw-wip-cbodley-testing---basic-mira/

**Pull request merged**